src/node/desktop/src/main/detect_r.ts

```diff
@@ -37,116 +42,98 @@ function showRNotFoundError(msg: string): void {
-  }
-}
 
-async function prepareEnvironmentPosix(): Promise<boolean> {
-
-  assert((process.platform !== 'win32'));
-
-  // check for which R override
-  let rWhichRPath = new FilePath();
-  const whichROverride = getenv('RSTUDIO_WHICH_R');
-  if (whichROverride) {
-    rWhichRPath = new FilePath(whichROverride);
+  try {
+    return prepareEnvironmentImpl();
+  } catch (error) {
+    logger().logError(error);
+    return false;
   }
 
-  let rLdScriptPath = new FilePath();
-  if (process.platform === 'darwin') {
-    if (app.isPackaged) {
-      rLdScriptPath = appState().scriptsPath?.completePath('session/r-ldpath') ?? new FilePath();
-      if (!rLdScriptPath.existsSync()) {
-        rLdScriptPath = new FilePath(app.getAppPath()).completePath('r-ldpath');
-      }
-    } else {
-      rLdScriptPath = appState().sessionPath?.completePath('../r-ldpath') ?? new FilePath();
-    }
-  } else {
-    // determine rLdPaths script location
-    rLdScriptPath = appState().supportingFilePath().completePath('bin/r-ldpath');
-    if (!rLdScriptPath.existsSync()) {
-      rLdScriptPath = appState().supportingFilePath().completePath('session/r-ldpath');
-    }
-  }
+}
+
+async function prepareEnvironmentImpl(): Promise<boolean> {
 
   // attempt to detect R environment
-  const detectResult = await detectREnvironment(rWhichRPath, rLdScriptPath, '');
-  if (!detectResult.success) {
-    showRNotFoundError(detectResult.errMsg ?? 'Unknown error');
+  let rEnvironment : REnvironment;
+  try {
+    rEnvironment = await detectREnvironment();
+  } catch (error) {
+    showRNotFoundError(error ?? 'An unknown error occurred.');
     return false;
   }
 
-  logger().logDiagnostic(`Using R script: ${detectResult.rScriptPath}`);
+  // set environment variables from R
+  setVars(rEnvironment.envVars);
 
-  setREnvironmentVars(detectResult.envVars ?? {});
-
-  if (process.platform === 'darwin') {
-    // TODO: deal with selection of x86_64 or arm64 build of R
+  // on Windows, ensure R is on the PATH so that companion DLLs
+  // in the same directory can be resolved
+  const scriptPath = rEnvironment.rScriptPath;
+  if (process.platform === 'win32') {
+    const binDir = dirname(scriptPath);
+    process.env.PATH = `${binDir};${process.env.PATH}`;
   }
 
   return true;
-}
 
-async function prepareEnvironmentWin32(): Promise<boolean> {
-  assert((process.platform === 'win32'));
+}
 
-  // TODO: this is just a hacked-together placeholder for dev purposes
+async function detectREnvironment(): Promise<REnvironment> {
 
-  // check for which R override
-  let rWhichRPath = new FilePath();
-  const whichROverride = getenv('RSTUDIO_WHICH_R');
-  if (!whichROverride) {
-    dialog.showErrorBox(
-      'RSTUDIO_WHICH_R Not Set',
-      'Temporary: RSTUDIO_WHICH_R environment variable must point to R installation folder. Cannot continue.');
-    return false;
+  // scan for R
+  const R = await scanForR();
+  if (!R) {
+    throw rNotFoundErrorMessage();
```

Comment:
Here, and all other locations where you throw rNotFoundErrorMessage: always throw an Error or subclass of an Error. So, `throw Error(rNotFoundErrorMessage())` would work.

From https://nodejs.org/dist/latest-v14.x/docs/api/errors.html:

> Class: Error
> A generic JavaScript <Error> object that does not denote any specific circumstance of why the error occurred. Error objects capture a "stack trace" detailing the point in the code at which the Error was instantiated, and may provide a text description of the error.
> 
> All errors generated by Node.js, including all system and JavaScript errors, will either be instances of, or inherit from, the Error class.
```